### PR TITLE
remote-write: respect Retry-After header on 5xx errors

### LINKED ISSUE #12674: Respect Retry-After header on non-429 requests.
### Proposal

The `retry_on_http_429` field on queue_config allows Prometheus to retry requests when the server responds with the matching status code. On the other hand, for `5xx` error codes, Prometheus will retry using the min_backoff/max_backoff parameters only, ignoring any Retry-After headers.

In cases that the remote endpoint responds with an informed header value, it will not be respected and an erroneous queue_configs can lead to overloading it.

Given that 5xx are understood to be retryable and can happen in different circumstances, I think it would be a good idea to allow Prometheus to respect the Retry-After headers on 5xx responses as a _minimum_ time of waiting and start the min_backoff approach _after_ this time has elapsed.

What do you think?

### PR DESCRIPTION
**Pull request by bboreham**
If the server sent it to us, we should assume it knows better than we do and respect it.

Fixes #12674

### CODE REVIEW

**Comment by cstyan** [comment]
:+1: fine by me, can we update a test similar to `TestClientRetryAfter` and docs as well stating that the header will be respected for 500 errors

**Comment by wildum** [comment]
I created a PR to @bboreham's branch with a refactor of the test and an additional test case for the HTTP 500 retry: https://github.com/bboreham/prometheus/pull/2

**Comment by bboreham** [comment]
I have included William's test (thanks!) and rebased.

**Review by bwplotka — APPROVED** [review]
Thanks 👍🏽

File: storage/remote/client.go [hunk 1]
```diff
@@ -236,10 +236,8 @@ func (c *Client) Store(ctx context.Context, req []byte, attempt int) error {
 		}
 		err = fmt.Errorf("server returned HTTP status %s: %s", httpResp.Status, line)
 	}
-	if httpResp.StatusCode/100 == 5 {
-		return RecoverableError{err, defaultBackoff}
-	}
-	if c.retryOnRateLimit && httpResp.StatusCode == http.StatusTooManyRequests {
+	if httpResp.StatusCode/100 == 5 ||
```
Comment (dmitrydprog) [inline storage/remote/client.go:239]: What is this???
Why not
`httpResp.StatusCode >= 500`

Comment (bboreham) [reply storage/remote/client.go:239]: For consistency with line 231.
It would be nice if the Go compiler recognized the common subexpressions.

**Pull request merged**